### PR TITLE
PaddlePaddle frontend - Core Part 

### PR DESCRIPTION
**Pull request by nosovmik**
### Details:
 - Basic classes for PDPD frontend
 - Test model generator approach for PDPD models testing
 - Sub-project for static 'libprotobuf' build (in addition to dynamic version used by onnx_importer)
 - Basic unit tests

### Tickets:
 - 52251, 52253, 52247, 55249, 55250, 54603, 52256

### CODE REVIEW

File: ngraph/frontend/cmake_static_protobuf/CMakeLists.txt
Comment (ilyachur) [inline ngraph/frontend/cmake_static_protobuf/CMakeLists.txt:19]: Why do you use relative path?
And why do you need to have static protobuf? Why cannot we use one protobuf for all OV components?

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/exceptions.hpp
Comment (ilyachur) [inline ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/exceptions.hpp:16]: Why do we need to have forward declaration here?

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/frontend.hpp
Comment (ilyachur) [inline ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/frontend.hpp:17]: Why do we need to have this private static method?

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/place.hpp
Comment (ilyachur) [inline ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/place.hpp:16]: Do you have cycle dependency for all classes for which you use forward declaration?

File: ngraph/CMakeLists.txt
Comment (nosovmik) [reply ngraph/CMakeLists.txt:316]: Now there are 2 versions of Protobuf: dynamic (for onnx import/export) and static (for PDPD, TF and others). Actually having one shared protobuf leads to various errors when loading/unloading plugins multiple times. Example is discussed here: https://stackoverflow.com/questions/37051635/several-shared-object-using-same-proto-leading-the-the-error-file-already-exist
And this error happens for our case when we create/destroy FrontEndManager.
Ideally it would be good to have only 'static' version of protobuf, however currently onnx_importer & onnx_editor have issues with that, they spend a lot of time to find necessary workarounds and came to 'dynamic' protobuf version. I decided to not spend a lot of time again to move 'onnx' part to static protobuf and create 'static' version for PDPD (and future TF) frontends.
I believe ONNX team will have issues with future 'onnx' frontend plugin, and they'll probably have to come back to 'static' version. On that time, we'll remove building of dynamic protobuf and keep only 'static' version

File: ngraph/frontend/cmake_static_protobuf/CMakeLists.txt
Comment (nosovmik) [reply ngraph/frontend/cmake_static_protobuf/CMakeLists.txt:19]: See my response in comment above

File: ngraph/CMakeLists.txt
Comment (tomdol) [reply ngraph/CMakeLists.txt:316]: Going back to static protobuf required by the onnx_importer will most probably bring back the problems caused by static version of protobuf. This will be a big regression.

Please test this PR against the ONNX importer unit tests and integration with ONNXRuntime(including windows) and make sure that the tests still pass and there are no segfaults.

@postrational FYI

Comment (nosovmik) [reply ngraph/CMakeLists.txt:316]: ONNX currently still uses 'dynamic protobuf', no changes are there. Only PDPD frontend uses static protobuf version and it doesn't conflict with any other components which uses dynamic protobuf. If we enable static protobuf for ONNX - we'll clearly get CI failures now, but this PR is green

Comment (mbencer) [reply ngraph/CMakeLists.txt:316]: >I believe ONNX team will have issues with future 'onnx' frontend plugin, and they'll probably have to come back to 'static' version. On that time, we'll remove building of dynamic protobuf and keep only 'static' version
1. As I know @slyalin do some experiments with onnx_importer/onnx_editor based on dynamic protobuf and it works correctly. Am I right?
2. As I understand in the current stack: we have dynamic protobuf used by onnx_importer and static by PDPD at the same time. Could you create a draft with `NGRAPH_PDPD_FRONTEND_ENABLE=ON` on all CI where `NGRAPH_ONNX_IMPORT_ENABLE=ON`? (probably it is worth to consider if some of them should have `NGRAPH_PDPD_FRONTEND_ENABLE=ON` permanently)

Comment (nosovmik) [reply ngraph/CMakeLists.txt:316]: > Could you create a draft with NGRAPH_PDPD_FRONTEND_ENABLE=ON on all CI where NGRAPH_ONNX_IMPORT_ENABLE=ON?
@mbencer Yes, NGRAPH_PDPD_FRONTEND_ENABLE is enabled in this PR everywhere where NGRAPH_ONNX_IMPORT_ENABLE=ON. See line 64. Also you can verify that PDPD frontend is enabled in CI 'cmake' logs (I added 'message' prints on line 116 in this file)

Comment (mbencer) [reply ngraph/CMakeLists.txt:316]: @nosovmik that's right, but some configurations are defined in https://github.com/openvinotoolkit/openvino/tree/master/.ci and some by external code (we can discuss offline). I've pushed for test: https://github.com/openvinotoolkit/openvino/pull/6274

File: ngraph/CMakeLists.txt
Comment (ilya-lavrenov) [inline ngraph/CMakeLists.txt:318]: have we checked that system protobuf has static libraries?

File: ngraph/frontend/CMakeLists.txt
Comment (ilya-lavrenov) [reply ngraph/frontend/CMakeLists.txt:19]: and `CMAKE_STATIC_LIBRARY_PREFIX`, `CMAKE_STATIC_LIBRARY_SUFFIX`

File: ngraph/frontend/paddlepaddle/CMakeLists.txt
Comment (ilya-lavrenov) [inline ngraph/frontend/paddlepaddle/CMakeLists.txt:16]: do we need it?

File: ngraph/frontend/paddlepaddle/CMakeLists.txt
Comment (ilya-lavrenov) [inline ngraph/frontend/paddlepaddle/CMakeLists.txt:99]: plugins don't need to install `ARCHIVE`, see https://github.com/openvinotoolkit/openvino/blob/master/cmake/developer_package/plugins/plugins.cmake#L125-L128

File: ngraph/frontend/paddlepaddle/CMakeLists.txt
Comment (ilya-lavrenov) [inline ngraph/frontend/paddlepaddle/CMakeLists.txt:65]: I suppose we need to use `MODULE` instead of `SHARED`? Since it's a plugin?

File: ngraph/test/CMakeLists.txt
Comment (ilya-lavrenov) [inline ngraph/test/CMakeLists.txt:509]: use `find_package(PythonInterp 3)`. 

Our CI and others pass `-DPYTHON_EXECUTABLE` which works only for `PythonInterp` while new cmkae module `Python3` does not take this value into account.

File: ngraph/CMakeLists.txt
Comment (nosovmik) [reply ngraph/CMakeLists.txt:318]: System protobuf has static libraries, but these are not used in a project. There is an option "NGRAPH_USE_SYSTEM_PROTOBUF", but it is always 'off' and enabling this produces ONNX build errors. So, I guess it is not used in any builds and probably should be removed under some cleanup activity. But I'd prefer to not do this in this PR

File: ngraph/frontend/paddlepaddle/CMakeLists.txt
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/CMakeLists.txt:16]: The original idea is to have PaddlePaddle frontend as a shared library, which can be potentially linked to user's applications. So that user app can cast 'InputModel' to 'PDPDInputModel' and call some paddlepaddle-specific functions
We'll check if this is still the case. If not, then I guess we can really try to implement this as 'module', not shared lib.
If yes, then public headers, archive, "shared library" settings should be kept as is

File: ngraph/test/CMakeLists.txt
Comment (nosovmik) [reply ngraph/test/CMakeLists.txt]: Fixed

**Review by ilya-lavrenov — APPROVED** [review]

File: ngraph/frontend/paddlepaddle/CMakeLists.txt
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/CMakeLists.txt:64]: We discussed internally and we'll continue using of "SHARED" library approach - integration with frameworks (PaddlePaddle) in our case will require work with concrete PDPDInputModel etc. At least it will be needed to define 'extensions' in future

File: ngraph/frontend/paddlepaddle/CMakeLists.txt
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/CMakeLists.txt:98]: Frontend is made as shared library, it is needed for integration with frameworks where concrete classes will be used

File: ngraph/frontend/paddlepaddle/CMakeLists.txt
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/CMakeLists.txt:15]: Discussed, decided to keep it as is (shared library, see other comments)

File: ngraph/frontend/cmake_static_protobuf/CMakeLists.txt
Comment (nosovmik) [reply ngraph/frontend/cmake_static_protobuf/CMakeLists.txt:19]: Regarding 'relative path' - static protobuf is currently planned to be used by frontends, thus it is placed inside "frontend" folder. It needs to include "external_protobuf.cmake" which is "ngraph/cmake" folder

File: ngraph/frontend/CMakeLists.txt
Comment (nosovmik) [reply ngraph/frontend/CMakeLists.txt]: Done. However was not able to use CMAKE_DEBUG_POSTFIX. This is because on linux it is empty (only on Windows it is 'd'), but protobuf has hardcoded its 'debug' name as libprotobufd for all platforms

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/exceptions.hpp
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/exceptions.hpp:16]: For me it is a good practice to keep in header files only this info which is required for inclusion. There is no need of full definition of NodeContext here, so not necessary to include appropriate header and forward declaration is enough.
Also it reduces compilation time.

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/frontend.hpp
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/frontend.hpp:17]: Convertion of model for PDPD frontend looks not dependent on any member variable, thus it was made static.
I guess it was just a "developer's preference", it could be 'non-static and const', but Frontend implementation for PDPD case doesn't have any class members

File: ngraph/frontend/paddlepaddle/src/frontend.cpp
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/src/frontend.cpp:48]: Removed

File: ngraph/frontend/paddlepaddle/CMakeLists.txt
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/CMakeLists.txt]: Removed commented code

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/place.hpp
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/place.hpp:16]: No, just full definition of this class is not needed for this header file to be compilable

File: ngraph/frontend/paddlepaddle/src/op/elementwise_ops.cpp
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/src/op/elementwise_ops.cpp:56]: Would it be ok if I create JIRAs to China team to resolve your 'op-specific' comments to speed up things? These operations are here just to be able to run unit tests to cover general PDPD code (as there is no way currently to define some 'mock' operations with 'mock protobuf').

File: ngraph/frontend/paddlepaddle/src/op/elementwise_ops.cpp
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/src/op/elementwise_ops.cpp:31]: Ticket 58539 is created for China team to handle this

File: ngraph/frontend/paddlepaddle/src/op/elementwise_ops.cpp
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/src/op/elementwise_ops.cpp:56]: Ticket 58539 is created for China team to handle this

File: ngraph/frontend/paddlepaddle/src/op/conv2d.cpp
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/src/op/conv2d.cpp:13]: Ticket 58539 is created for China team to handle this

File: ngraph/frontend/paddlepaddle/src/op/conv2d_utils.cpp
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/src/op/conv2d_utils.cpp:104]: Ticket 58539 is created for China team to handle this

File: ngraph/CMakeLists.txt
Comment (ilyachur) [reply ngraph/CMakeLists.txt:318]: @mbencer @tomdol Do you have any objections against the current solution? Can we get conflicts?

**Review by jane-intel — APPROVED** [review]

File: ngraph/frontend/paddlepaddle/src/op/elementwise_ops.cpp
Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/elementwise_ops.cpp:56]: @liubo-intel  Please have a try to use Unsqueeze

File: ngraph/frontend/paddlepaddle/src/op/elementwise_ops.cpp
Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/elementwise_ops.cpp:31]: @liubo-intel Please have an investigation

File: ngraph/frontend/paddlepaddle/src/op/conv2d.cpp
Comment (zhangYiIntel) [reply ngraph/frontend/paddlepaddle/src/op/conv2d.cpp:13]: Hi Jane, currently PDPD doesn't use opset version to distinguish different versions of op. Instead, they use different name for different versions of the ops, for example, they have `nearest_interpolate` and `nearest_interpolate_v2`.

File: ngraph/CMakeLists.txt
Comment (ilyachur) [reply ngraph/CMakeLists.txt:318]: @mbencer @tomdol Guys, could you please comment?

Comment (tomdol) [reply ngraph/CMakeLists.txt:318]: Sorry, I missed the GH notification. I still think that going back to the static PB in the future will bring the previous problems back. According to this answer https://stackoverflow.com/a/37056328/4517806 it's possible to use static protobuf but each library will have to be linked with a copy of PB. This will increase the size of each library and will require some extra serialization/deserialization pairs. For example The model will have to be deserialized in the ONNX editor, modified, serialized to a string so that it can be safely passed to the importer, deserialized in the latter and only then converted to an ngraph function. I'm still not 100% sure that will work.

I definitely don't want to block this PR and my approval is not required here. I suggest consulting it with @postrational because he's in charge of the FE API integration for the onnx_importer.

Comment (ilyachur) [reply ngraph/CMakeLists.txt:320]: Ok, I propose to return back to this discussion if we have any issues with ONNX Frontend.

File: ngraph/test/CMakeLists.txt
Comment (mbencer) [inline ngraph/test/CMakeLists.txt:522]: Shouldn't we use here at first python passed by `-DPYTHON_EXECUTABLE=`?

File: ngraph/CMakeLists.txt
Comment (mbencer) [reply ngraph/CMakeLists.txt:320]: Based on CI results (#6274) having two instances of protobuf (shared for onnx and static for pdpd) doesn't cause problems on any tested platform.
I also downloaded this branch and prepare unit tests where both pdpd and onnx_importer are used in the same process. At least on my dev machine, I haven't observed problems (we should have such tests in CI after ONNXImporter/FrontendApi integration).
The code with double protobuf is not very elegant, but I don't know how to deal with it now.

**Review by mbencer — CHANGES_REQUESTED** [review]

File: ngraph/CMakeLists.txt
Comment (mbencer) [reply ngraph/CMakeLists.txt:320]: Based on CI results (#6274) having two instances of protobuf (shared for onnx and static for pdpd) doesn't cause problems on any tested platform.
I also downloaded this branch and prepare unit tests where both pdpd and onnx_importer are used in the same process. At least on my dev machine, I haven't observed problems (we should have such tests in CI after ONNXImporter/FrontendApi integration).
The code with double protobuf is not very elegant, but I don't know how to deal with it now.

File: ngraph/test/CMakeLists.txt
Comment (nosovmik) [reply ngraph/test/CMakeLists.txt:522]: This way should take -DPYTHON_EXECUTABLE correctly, please see this discussion https://github.com/openvinotoolkit/openvino/pull/6036#discussion_r656008155

Comment (mbencer) [reply ngraph/test/CMakeLists.txt:522]: Ok, I see, I've tested locally the version before this fix

Comment (nosovmik) [reply ngraph/test/CMakeLists.txt:522]: @mbencer great, can you please approve this PR then? You've "requested changes" and merge will be blocked without your approval

**Review by mbencer — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

File: ngraph/frontend/paddlepaddle/src/op/elementwise_ops.cpp
Comment (ceciliapeng2011) [reply ngraph/frontend/paddlepaddle/src/op/elementwise_ops.cpp:31]: @jane-intel There will only be static rank to inputs.  Confirmed with Baidu PaddleLite arch and Paddle op engineers. So let's defer this rank relax implementation. It will also make the converted nGraph function simple and friend to performance.

**Pull request merged**
